### PR TITLE
11065 ess start downloading enc file from fss as soon as possible

### PR DESCRIPTION
**Pull request by Siddartha-tech**
Coding changes.
Updated Test Case

### CODE REVIEW

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs [hunk 1]
```diff
@@ -238,32 +243,50 @@ private async Task<string> SelectLatestPublishedDateBatch(List<Products> product
...
             if (matchEditionNumber.Any(a => a.Value == productItem.Cancellation.EditionNumber.Value.ToString()))
             {
                 matchEditionNumber.ForEach(c => c.Value = Convert.ToString(productItem.EditionNumber));
                 internalSearchBatchResponse.Entries.Add(item);
                 productList.Add(compareProducts);
+                await StartPreparingForDownload(item, productItem, exchangeSetRootPath, message, cancellationTokenSource, cancellationToken);
             }
         }
 
+        private async Task StartPreparingForDownload(BatchDetail item, Products productItem, string exchangeSetRootPath, SalesCatalogueServiceResponseQueueMessage message, CancellationTokenSource cancellationTokenSource, CancellationToken cancellationToken)
```
Comment (barrie-cooper) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs:279]: Is this method named right?  Seems to be more of a Download than a Prepare.  Maybe `PerformDownload` or `PerformBatchFileDownload`?

Comment (Siddartha-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs]: Updated method name. Thanks!

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs [hunk 2]
```diff
@@ -118,16 +118,9 @@ public async Task<SalesCatalogueProductResponse> DownloadSalesCatalogueResponse(
 
         public async Task<List<FulfilmentDataResponse>> QueryAndDownloadFileShareServiceFiles(SalesCatalogueServiceResponseQueueMessage message, List<Products> products, string exchangeSetRootPath, CancellationTokenSource cancellationTokenSource, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs]: Updated method name. Thanks!

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs [hunk 3]
```diff
@@ -238,32 +243,50 @@ private async Task<string> SelectLatestPublishedDateBatch(List<Products> product
...
+        {
+            var ProductName = productItem.ProductName;
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs]: Updated variable name. And Unit Test case is already is there only thing is calling changes to direct within the FileShareService class. As earlier it was not. Means it was called from FulfilmentDataService class.
Thanks!

**Review by barrie-cooper — APPROVED** [review]
LGTM

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs [hunk 4]
```diff
@@ -119,35 +119,35 @@ public async Task<SearchBatchResponse> GetBatchInfoBasedOnProducts(List<Products
...
+                    logger.LogError(EventIds.QueryFileShareServiceENCFilesNonOkResponse.ToEventId(), "Error in file share service while searching and downloading ENC files with uri:{RequestUri}, responded with {StatusCode} and BatchId:{batchId} and _X-Correlation-ID:{correlationId}", httpResponse.RequestMessage.RequestUri, httpResponse.StatusCode, message.BatchId, message.CorrelationId);
+                    logger.LogError(EventIds.CancellationTokenEvent.ToEventId(), "Request cancelled for Error in file share service while searching and downloading ENC files with cancellationToken:{cancellationTokenSource.Token} with uri:{RequestUri}, responded with {StatusCode} and BatchId:{batchId} and _X-Correlation-ID:{correlationId}", JsonConvert.SerializeObject(cancellationTokenSource.Token), httpResponse.RequestMessage.RequestUri, httpResponse.StatusCode, message.BatchId, message.CorrelationId);
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs:142]: Please remove "and downloading" from both logs as this response is for search only.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs [hunk 5]
```diff
@@ -119,35 +119,35 @@ public async Task<SearchBatchResponse> GetBatchInfoBasedOnProducts(List<Products
             var prodCount = products.Select(a => a.UpdateNumbers).Sum(a => a.Count);
             int queryCount = 0;
 
-            logger.LogInformation(EventIds.FileShareServicePreparingToSearchSetOfENCsStarted.ToEventId(), "Preparing file share service search request for {productDetails}. ESS BatchId:{batchId} and _X-Correlation-ID:{CorrelationId}", productWithAttributes.Item2, batchId, correlationId);
+            logger.LogInformation(EventIds.FileShareServicePreparingToSearchSetOfENCsStarted.ToEventId(), "Preparing file share service search and download request for {productDetails}. ESS BatchId:{batchId} and _X-Correlation-ID:{CorrelationId}", productWithAttributes.Item2, message.BatchId, message.CorrelationId);
```
[the remark this answers is not in the document]
Comment (Siddartha-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs:122]: Did changes accordingly. Thanks!

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs (hunk 4 above)
Comment (Siddartha-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs:142]: Did changes accordingly. Thanks!

**Review by janvi13603 — APPROVED** [review]
Thanks for fixing review comments.

**Comment by barrie-cooper** [comment]
Trying to merge this and there appear to be 4 failing unit tests.  Could you have a look please?

https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_build/results?buildId=24596&view=logs&j=30945a94-8ee8-5fb4-183e-8cd5348ef8d1&t=584f202c-0fab-5bcf-25b6-8c9e2a6cdbdb

@Siddartha-tech @janvi13603

**Pull request merged**